File: typescript/serverless-food-ordering/app/src/services/order_service.ts
import * as restate from "@restatedev/restate-sdk";
import * as delivery from "./delivery";
import { v4 as uuidv4 } from "uuid";
import {Order, Status} from "../types/types";
import * as orderstatus from "./order_status_service";
import { getStripeClient } from "../clients/stripe_client";
import { completed } from "../utils/utils";
import * as restaurantProxy from "./restaurant_proxy";

const stripe = getStripeClient();

const eventHandler = async (ctx: restate.RpcContext, event: restate.Event) => {
  const order = JSON.parse(event.json<string>());
  await create(ctx, event.key, order);
};

const create = async (
  ctx: restate.RpcContext,
  orderId: string,
  order: Order
) => {
  // Ignore already completed orders
  if (await completed(ctx)) {
    return
  }

  const { id, totalCost, deliveryDelay } = order;

  // 1. Set status
  ctx.send(orderstatus.service).setStatus(id, Status.CREATED);

  // 2. Handle payment
  const token = await ctx.sideEffect(async () => uuidv4());
  const paid = await ctx.sideEffect(() => stripe.charge(id, token, totalCost));

  if (!paid) {
    ctx.send(orderstatus.service).setStatus(id, Status.REJECTED);
    return;
  }

  // 3. Schedule preparation
  ctx.send(orderstatus.service).setStatus(id, Status.SCHEDULED);
  await ctx.sleep(deliveryDelay);

  // 4. Trigger preparation
  const preparationPromise = ctx.awakeable();
  // The locally running POS server is not reachable by Lambda, so to simplify the setup the restaurant is also a Restate service.
  // If the POS server would be deployed somewhere, this would be a side effect that sends an HTTP request to the restaurant.
  ctx.send(restaurantProxy.service).prepare(id, preparationPromise.id);
  ctx.send(orderstatus.service).setStatus(id, Status.IN_PREPARATION);

  await preparationPromise.promise;
  ctx.send(orderstatus.service).setStatus(id, Status.SCHEDULING_DELIVERY);

  // 5. Find a driver and start delivery
  await delivery.createDelivery(ctx, order);
  ctx.send(orderstatus.service).setStatus(id, Status.DELIVERED);
};

export const router = restate.keyedRouter({
  create,
  eventHandler: restate.keyedEventHandler(eventHandler),
});

export type orderApi = typeof router;
export const service: restate.ServiceApi<orderApi> = { path: "orders" };

File: java/food-ordering/app/restate-app/src/main/java/dev/restate/sdk/examples/OrderService.java
package dev.restate.sdk.examples;

import static dev.restate.sdk.examples.utils.TypeUtils.statusToProto;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import dev.restate.sdk.RestateContext;
import dev.restate.sdk.common.CoreSerdes;
import dev.restate.sdk.common.TerminalException;
import dev.restate.sdk.examples.clients.PaymentClient;
import dev.restate.sdk.examples.clients.RestaurantClient;
import dev.restate.sdk.examples.generated.DeliveryServiceRestate;
import dev.restate.sdk.examples.generated.OrderProto;
import dev.restate.sdk.examples.generated.OrderServiceRestate;
import dev.restate.sdk.examples.generated.OrderStatusServiceRestate;
import dev.restate.sdk.examples.types.OrderRequest;
import dev.restate.sdk.examples.types.Status;
import java.time.Duration;
import java.util.UUID;

public class OrderService extends OrderServiceRestate.OrderServiceRestateImplBase {
  private final RestaurantClient restaurant = RestaurantClient.get();
  private final PaymentClient paymentClnt = PaymentClient.get();

  @Override
  public void handleOrderCreationEvent(RestateContext ctx, OrderProto.KafkaOrderEvent event)
      throws TerminalException {
    var orderStatusSend = OrderStatusServiceRestate.newClient(ctx);

    try {
      ObjectMapper mapper = new ObjectMapper();
      OrderRequest order = mapper.readValue(event.getPayload().toStringUtf8(), OrderRequest.class);

      // 1. Set status
      orderStatusSend.oneWay().setStatus(statusToProto(order.orderId, Status.CREATED));

      // 2. Handle payment
      String token = ctx.sideEffect(CoreSerdes.STRING_UTF8, () -> UUID.randomUUID().toString());
      boolean paid =
          ctx.sideEffect(
              CoreSerdes.BOOLEAN, () -> paymentClnt.charge(order.orderId, token, order.totalCost));

      if (!paid) {
        orderStatusSend.oneWay().setStatus(statusToProto(order.orderId, Status.REJECTED));
        return;
      }

      // 3. Schedule preparation
      orderStatusSend.setStatus(statusToProto(order.orderId, Status.SCHEDULED));
      ctx.sleep(Duration.ofMillis(order.deliveryDelay));

      // 4. Trigger preparation
      var preparationAwakeable = ctx.awakeable(CoreSerdes.VOID);
      ctx.sideEffect(() -> restaurant.prepare(order.orderId, preparationAwakeable.id()));
      orderStatusSend.setStatus(statusToProto(order.orderId, Status.IN_PREPARATION));

      preparationAwakeable.await();
      orderStatusSend.setStatus(statusToProto(order.orderId, Status.SCHEDULING_DELIVERY));

      // 5. Find a driver and start delivery
      var deliveryAwakeable = ctx.awakeable(CoreSerdes.VOID);

      var deliveryRequest =
          OrderProto.DeliveryRequest.newBuilder()
              .setOrderId(order.orderId)
              .setRestaurantId(order.restaurantId)
              .setCallback(deliveryAwakeable.id())
              .build();
      DeliveryServiceRestate.newClient(ctx).oneWay().start(deliveryRequest);
      deliveryAwakeable.await();
      orderStatusSend.setStatus(statusToProto(order.orderId, Status.DELIVERED));
    } catch (JsonProcessingException e) {
      throw new TerminalException("Parsing raw JSON order failed: " + e.getMessage());
    }
  }
}
